src/main/java/io/reactivex/internal/operators/flowable/FlowableLimit.java

```diff
@@ -0,0 +1,137 @@
+/**
+ * Copyright (c) 2016-present, RxJava Contributors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in
+ * compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is
+ * distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See
+ * the License for the specific language governing permissions and limitations under the License.
+ */
+
+package io.reactivex.internal.operators.flowable;
+
+import java.util.concurrent.atomic.AtomicLong;
+
+import org.reactivestreams.*;
+
+import io.reactivex.*;
+import io.reactivex.annotations.Experimental;
+import io.reactivex.internal.subscriptions.*;
+import io.reactivex.plugins.RxJavaPlugins;
+
+/**
+ * Limits both the total request amount and items received from the upstream.
+ *
+ * @param <T> the source and output value type
+ * @since 2.1.6 - experimental
+ */
+@Experimental
+public final class FlowableLimit<T> extends AbstractFlowableWithUpstream<T, T> {
+
+    final long n;
+
+    public FlowableLimit(Flowable<T> source, long n) {
+        super(source);
+        this.n = n;
+    }
+
+    @Override
+    protected void subscribeActual(Subscriber<? super T> s) {
+        source.subscribe(new LimitSubscriber<T>(s, n));
+    }
+
+    static final class LimitSubscriber<T>
+    extends AtomicLong
+    implements FlowableSubscriber<T>, Subscription {
+
+        private static final long serialVersionUID = 2288246011222124525L;
+
+        final Subscriber<? super T> actual;
+
+        long remaining;
+
+        Subscription upstream;
+
+        LimitSubscriber(Subscriber<? super T> actual, long remaining) {
+            this.actual = actual;
+            this.remaining = remaining;
+            lazySet(remaining);
+        }
+
+        @Override
+        public void onSubscribe(Subscription s) {
+            if (SubscriptionHelper.validate(this.upstream, s)) {
+                if (remaining == 0L) {
+                    s.cancel();
+                    EmptySubscription.complete(actual);
+                } else {
+                    this.upstream = s;
+                    actual.onSubscribe(this);
+                }
+            }
+        }
+
+        @Override
+        public void onNext(T t) {
+            long r = remaining;
+            if (r > 0L) {
+                remaining = --r;
+                actual.onNext(t);
+                if (r == 0L) {
+                    upstream.cancel();
```

Comment:
What about this:

```java
if (r > 0L) {
  remaining = --r;
  final boolean completed = r == 0L;
  
  if (completed) {
    upstream.cancel();
  }

  actual.onNext(t);

  if (completed) {
    actual.onComplete();
  }
}
```

This way we cancel upstream before delivering `onNext` notification, it could be important in case if `onNext` triggers long running chain of computations thus delaying cancellation of the upstream (in current version).

Reply:
Though, current implementations of `take()` also does this so it might be unwanted behavior change, up2u, we can move this change to 3.x

Reply:
Or it could trigger an interrupted thread and fail code in the downstream unexpectedly.

Reply:
Good point! 

But doesn't it mean that `actual.onComplete()` should happen before `upstream.cancel()` then? :trollface:

Reply:
It's way less likely to happen. There were issues around `onNext` before.

Reply:
ok, although I think probability is the same here, `upstream.cancel()` has to be called in both ways of call ordering and possibly cause problems. But in current order we guarantee delivery of onNext which is probably good :)